sklearn/feature_extraction/dict_vectorizer.py

```diff
@@ -162,9 +165,14 @@ def _transform(self, X, fitting):
 
         # collect all the possible feature names and build sparse matrix at
         # same time
+        if self.drop_first_category:
+            to_drop = {}
         for x in X:
             for f, v in six.iteritems(x):
                 if isinstance(v, six.string_types):
+                    if self.drop_first_category and f not in to_drop:
```

Comment:
It's a pity we still add this effort every sample for every feature. I also fear this is too sensitive to the ordering of the data for the user to find it explicable.

We could drop the alphanumerically first entry by using a find-min. Something like:

```python
to_drop = {}
k, v in vocab.items():
    prefix = k.rpartition(sep)[0]
    if prefix not in to_drop or k < to_drop[prefix]:
        to_drop[prefix] = k
```
We do encounter problems of potential amibugity in `sep`.

You need to update the vocab indices below too.